cardano-api/src/Cardano/Api/SerialiseLedgerCddl.hs

```diff
@@ -0,0 +1,318 @@
+import           Cardano.Api.Error
+import           Cardano.Api.HasTypeProxy
+import           Cardano.Api.SerialiseCBOR
+import           Cardano.Api.Tx
+
+
+-- Why have we gone this route? The serialization format of `TxBody era`
+-- differs from the CDDL. We serialize to an intermediate type in order to simplify
+-- the specification of Plutus scripts and to avoid users having to think about
+-- and construct redeemer pointers. However it turns out we can still serialize to
+-- the ledger's CDDL format and maintain the convenient script witness specification
+-- that the cli commands build and build-raw expose.
+--
+-- The long term plan is to have all relevant outputs from the cli to adhere to
+-- the ledger's CDDL spec. Modifying the existing TextEnvelope machinery to encompass
+-- this would result in a lot of unnecessary changes where the serialization
+-- already defaults to the CDDL spec. In order to reduce the number of changes, and to
+-- ease removal of the non-CDDL spec serialization, we have opted to create a separate
+-- data type to encompass this in the interim.
+
+data TextEnvelopeCddl = TextEnvelopeCddl
+  { teCddlType :: !Text
+  , teCddlDescription :: !Text
+  , teCddlRawCBOR :: !ByteString
+  } deriving (Eq, Show)
+
+instance ToJSON TextEnvelopeCddl where
+  toJSON TextEnvelopeCddl {teCddlType, teCddlDescription, teCddlRawCBOR} =
+    object [ "type"        .= teCddlType
+           , "description" .= teCddlDescription
+           , "cborHex"     .= Text.decodeUtf8 (Base16.encode teCddlRawCBOR)
+           ]
+
+instance FromJSON TextEnvelopeCddl where
+  parseJSON = withObject "TextEnvelopeCddl" $ \v ->
+                TextEnvelopeCddl <$> (v .: "type")
+                                 <*> (v .: "description")
+                                 <*> (parseJSONBase16 =<< v .: "cborHex")
+    where
+      parseJSONBase16 v =
+        either fail return . Base16.decode . Text.encodeUtf8 =<< parseJSON v
+
+
+data TextEnvelopeCddlError
+  = TextEnvelopeCddlErrCBORDecodingError DecoderError
+  | TextEnvelopeCddlAesonDecodeError FilePath String
+  | TextEnvelopeCddlUnknownKeyWitness
+  | TextEnvelopeCddlTypeError
+      [Text] -- ^ Expected types
+      Text   -- ^ Actual types
+  | TextEnvelopeCddlErrUnknownType Text
+  | TextEnvelopeCddlErrByronKeyWitnessUnsupported
+  deriving (Show, Eq)
+
+instance Error TextEnvelopeCddlError where
+  displayError (TextEnvelopeCddlErrCBORDecodingError decoderError) =
+    "TextEnvelopeCDDL CBOR decoding error: " <> show decoderError
+  displayError (TextEnvelopeCddlAesonDecodeError fp aesonErr) =
+    "Could not JSON decode TextEnvelopeCddl file at: " <> fp <> " Error: " <> aesonErr
+  displayError TextEnvelopeCddlUnknownKeyWitness =
+    "Unknown key witness specified"
+  displayError (TextEnvelopeCddlTypeError expTypes actType) =
+    "TextEnvelopeCddl type error: "
+       <> " Expected one of: "
+       <> List.intercalate ", "
+            [Text.unpack expType | expType <- expTypes]
```

Comment:
Can this be replaced with `fmap Text.unpack expTypes`?

Reply:
Yep, this was a copy paste.